### PR TITLE
[dagstermill] save notebook on failure

### PR DESCRIPTION
**Pull request by jamiedemaria**
### Summary & Motivation
Adds the option to save the notebook to a storage location if the notebook fails during execution

### How I Tested These Changes
new unit tests + manual test in the example repository (materialize the error notebook asset and then confirm that i can open the saved notebook)

### CODE REVIEW

File: python_modules/libraries/dagstermill/dagstermill/asset_factory.py [hunk 1]
```diff
@@ -122,6 +135,7 @@ def define_dagstermill_asset(
     group_name: Optional[str] = None,
     io_manager_key: Optional[str] = None,
     retry_policy: Optional[RetryPolicy] = None,
+    save_notebook_on_failure: bool = False,
```
Comment (yuhan) [inline python_modules/libraries/dagstermill/dagstermill/asset_factory.py:138]: should this be `Optional[bool]` instead?

File: python_modules/libraries/dagstermill/dagstermill/asset_factory.py [hunk 2]
```diff
@@ -238,5 +255,6 @@ def iris_dataset():
         _dm_compute(
             name=name,
             notebook_path=notebook_path,
+            save_notebook_on_failure=save_notebook_on_failure,
```
Comment (yuhan) [inline python_modules/libraries/dagstermill/dagstermill/asset_factory.py:258]: prob wanna capture the invalid type early on by `save_notebook_on_failure=check.opt_bool_param(save_notebook_on_failure)`

File: python_modules/libraries/dagstermill/dagstermill/asset_factory.py (hunk 1 above)
Comment (jamiedemaria) [reply python_modules/libraries/dagstermill/dagstermill/asset_factory.py:138]: what are you thinking would be the behavior if it's None?

Comment (yuhan) [reply python_modules/libraries/dagstermill/dagstermill/asset_factory.py:155]: `opt_bool_param` should convert None to false by default

Comment (yuhan) [reply python_modules/libraries/dagstermill/dagstermill/asset_factory.py:155]: so the idea is user-facing wise, they can omit this arg but once we validate user inputs, all the args to `_dm_compute` should be not null.

File: python_modules/libraries/dagstermill/dagstermill/asset_factory.py [hunk 3]
```diff
@@ -205,6 +222,7 @@ def iris_dataset():
 
     check.str_param(name, "name")
     check.str_param(notebook_path, "notebook_path")
+    check.opt_bool_param(save_notebook_on_failure, "save_notebook_on_failure")
```
Comment (yuhan) [inline python_modules/libraries/dagstermill/dagstermill/asset_factory.py:225]: can do `save_notebook_on_failure = check.opt_bool_param(save_notebook_on_failure, "save_notebook_on_failure")` so when `save_notebook_on_failure` is not provided by the user, we fill it in as `false`

File: python_modules/libraries/dagstermill/dagstermill/asset_factory.py (hunk 1 above)
Comment (yuhan) [reply python_modules/libraries/dagstermill/dagstermill/asset_factory.py:155]: hmm i was mistaken `bool = False` and `Optional[bool] = False` both would work.

**Review by yuhan — APPROVED** [review]

**Pull request merged**